utils.py

```diff
@@ -0,0 +1,243 @@
+
+    Parameters
+    ----------
+    file : TYPE str 
+        DESCRIPTION path to a text file with one column whose rows are converted to list elements
+    sep : TYPE str
+        DESCRIPTION the separator for each row. Typically new line (default)
+
+    Returns
+    -------
+    Python list with elements corresponding to the rows of file
+
+    """
+    
+    f = open(file, 'r')
+    r = [line.split(sep)[0] for line in f]
+    return r
+
+def check_spatial_segs(segs, sep='\n'):
+    """
+    Checks if segs is None or a file name. 
+    If a file name, it uses txt_to_list to convert to a list and returns the list.
+
+    Parameters
+    ----------
+    segs : TYPE str
+        DESCRIPTION 'None' or the path to a text file that is used in txt_to_list
+    sep : TYPE str
+        DESCRIPTION the separator for each row. Typically new line (default)
+
+    Returns
+    -------
+    None or a Python list with elements corresponding to the rows of the segs file
+
+    """
+    
+    if segs != 'None':
+        r = txt_to_list(segs, sep)
+    else:
+        r = None
+    return r
+
+def write_preds_obs(pred_file, obs_file, partition, spatial_idx_name, 
+                    time_idx_name, filepath, spatial=False, train_sites=None, 
+                    val_sites=None, test_sites=None):
+    """
+    Joins the predictions and observations, and writes the result to a csv file
+    
+    Parameters
+    ----------
+    pred_file : TYPE str
+        DESCRIPTION filepath to the predictions file
+    obs : TYPE str
+        DESCRIPTION filepath to the observations file
+    partition : TYPE str
+        DESCRIPTION one of 'trn', 'val', or 'tst'
+    spatial_idx_name : TYPE str 
+        DESCRIPTION name of column that is used for spatial
+            index (e.g., 'seg_id_nat')
+    time_idx_name : TYPE str
+        DESCRIPTION name of column that is used for temporal index
+            (usually 'time')
+    filepath : TYPE str
+        DESCRIPTION path name of the output file
+    spatial : TYPE bool
+        DESCRIPTION when True, the pred_file is trimmed to the reaches according
+        to each of the provided splits
+    val_sites : TYPE list 
+        sites to exclude from training and test metrics
+    test_sites : TYPE list 
+        sites to exclude from validation and training metrics
+    train_sites : TYPE list 
+        sites to exclude from validation and test metrics
+
+    Returns
+    -------
+    None.
+    Writes a file with columns for the time_idx_name, spatial_idx_name, obs, and pred
+    """
+    var_data = fmt_preds_obs(pred_file, obs_file, 
+                             spatial_idx_name, time_idx_name)
+
+    for data_var, data in var_data.items():
+        #reset the index so that the time and space indicies are attributes
+        data.reset_index(inplace=True)
+        
+        if spatial:
+            # mask out validation and test sites from trn partition
+            if train_sites and partition == 'trn':

```

Comment:
To confirm, the purpose of including the `List` (i.e., `{partition}_sites`) in these checks is because they will fail if the `List` is empty, right?

Context: I originally thought the only thing that seemed to matter was if `partition` was the specified string.

Reply:
Yes, an empty `List` seems to have the same behavior as `False` within an if statement. So, providing the `List` here is a check to see if any train/val/test sites are specified